magma/magma
# [feg][aaa][radius] Add support for Radius Dynamic Authorization Extensions (RFC5176) to AAA radius server

Signed-off-by: Evgeniy Makeev <evgeniym@fb.com>

## Summary

Added support for Radius Dynamic Authorization Extensions to builtin AAA Server radius. To enable builtin radius 'with_builtin_radius' is required. The default build creates legacy aaa server which relies on an external radius server (see magma/feg/radius).

## Test Plan

FeG: make precommit;

build & deploy on test OpenWRT AP:
GOOS=linux GOARCH=arm go build -tags link_local_service,with_builtin_radius -o ~/bin/arm/ magma/feg/gateway/services/aaa/aaa_server
scp ~/bin/arm/aaa_server root@192.168.1.1:/sbin/

1) Start server and attach/detach user:

Mon Jul 27 18:44:45 2020 daemon.info aaa_server[14760]: I0728 01:44:45.297077   14760 in_memory.go:147] setting timeout of 300.000000 seconds for session: 3CC5B1AEE81A823D
Mon Jul 27 18:44:45 2020 daemon.info aaa_server[14760]: I0728 01:44:45.306230   14760 auth.go:161] successfully authenticated user: 0001010000000119@wlan.mnc001.mcc001.3gppnetwork.org
Mon Jul 27 18:44:45 2020 daemon.notice hostapd: wlan0: CTRL-EVENT-EAP-SUCCESS2 2e:73:09:75:e2:0d
Mon Jul 27 18:44:45 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d WPA: pairwise key handshake completed (RSN)
Mon Jul 27 18:44:45 2020 daemon.notice hostapd: wlan0: AP-STA-CONNECTED 2e:73:09:75:e2:0d
Mon Jul 27 18:44:45 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d RADIUS: starting accounting session 3CC5B1AEE81A823D
Mon Jul 27 18:44:45 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d IEEE 802.1X: authenticated - EAP type: 23 (unknown)
Mon Jul 27 18:44:45 2020 daemon.info aaa_server[14760]: I0728 01:44:45.326227   14760 accounting.go:81] setting timeout of 300.000000 seconds for session: 3CC5B1AEE81A823D
Mon Jul 27 18:44:45 2020 daemon.info aaa_server[14760]: I0728 01:44:45.326638   14760 acct.go:143] successfully handled Start Accounting Request for session_id:"3CC5B1AEE81A823D" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D"
...
Mon Jul 27 18:45:01 2020 daemon.notice hostapd: wlan0: AP-STA-DISCONNECTED 2e:73:09:75:e2:0d
Mon Jul 27 18:45:01 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d RADIUS: stopped accounting session 3CC5B1AEE81A823D
Mon Jul 27 18:45:01 2020 daemon.info aaa_server[14760]: I0728 01:45:01.727215   14760 acct.go:163] Interim Update: octets_in:9104 octets_out:7057 packets_in:85 packets_out:45 ctx:<session_id:"3CC5B1AEE81A823D" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D" >
Mon Jul 27 18:45:01 2020 daemon.info aaa_server[14760]: I0728 01:45:01.938095   14760 acct.go:143] successfully handled Stop Accounting Request for session_id:"3CC5B1AEE81A823D" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D"

2) Set session timeout to 30s, attach user & wait 4 session timeout to trigger DAE disconnect:

Mon Jul 27 18:54:45 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d IEEE 802.1X: authenticated - EAP type: 23 (unknown)
Mon Jul 27 18:54:45 2020 daemon.info aaa_server[15013]: I0728 01:54:45.385586   15013 accounting.go:81] setting timeout of 30.000000 seconds for session: 94867A6D4C610393
Mon Jul 27 18:54:45 2020 daemon.info aaa_server[15013]: I0728 01:54:45.385847   15013 acct.go:143] successfully handled Start Accounting Request for session_id:"94867A6D4C610393" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D"
Mon Jul 27 18:55:15 2020 daemon.info aaa_server[15013]: I0728 01:55:15.393003   15013 accounting.go:224] AAA session timeout for SID: 94867A6D4C610393, IMSI: 001010000000119
Mon Jul 27 18:55:15 2020 daemon.notice hostapd: wlan0: AP-STA-DISCONNECTED 2e:73:09:75:e2:0d
Mon Jul 27 18:55:15 2020 daemon.info hostapd: wlan0: STA 2e:73:09:75:e2:0d RADIUS: stopped accounting session 94867A6D4C610393
Mon Jul 27 18:55:15 2020 daemon.info aaa_server[15013]: I0728 01:55:15.607805   15013 acct.go:163] Interim Update: octets_in:8136 octets_out:6253 packets_in:65 packets_out:42 ctx:<session_id:"94867A6D4C610393" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D" >
Mon Jul 27 18:55:15 2020 daemon.warn aaa_server[15013]: W0728 01:55:15.608638   15013 accounting.go:116] Accounting Stop: Session 94867A6D4C610393 is not found
Mon Jul 27 18:55:15 2020 daemon.info aaa_server[15013]: I0728 01:55:15.608904   15013 acct.go:143] successfully handled Stop Accounting Request for session_id:"94867A6D4C610393" apn:"C6-41-1E-21-EF-55:OpenWrt_MDO" mac_addr:"2E-73-09-75-E2-0D"
Mon Jul 27 18:55:15 2020 daemon.info aaa_server[15013]: I0728 01:55:15.609923   15013 dae_builtin.go:60] DAE Disconnect Response Code Disconnect-ACK

## Additional Information
